lib/bundler/installer/gem_installer.rb

```diff
@@ -42,6 +42,7 @@ def failure_message
     end
 
     def install_error_message
+      Bundler.metrics.record_failed_install(spec.name, spec.version.to_s) unless Bundler.opt_out
```

Comment:
I'd move the `opt_out` check to the internals of Metrics class so that you don't need to check on every call.

Reply:
How would you implement it? I might be missing something but the way I see it, I'd still need to check on every call within all `metrics.rb` public methods with something like `return unless Bundler.opt_out`.
I'd also need to keep a `Metrics` object alive, which isn't required now.

Reply:
Yes, you'd still have to check on every Metrics public methods, but there are just four methods and it would replace many more (25?) checks. And I think it would be possible to code without requiring an instance of Metrics object if you use class variables.

Reply:
Hmm, interesting idea.
I'll have to change all instance variables to class variables, and all instance methods to class methods.
That's a significant design change, think it would be better?

Reply:
I wouldn't change everything just to not create an instance of `Metrics`, at least not until we are sure it must be done. It seems a premature optimization. I'd only move the opt_out checks. I wouldn't mind about `start` either.

Reply:
I've already gone ahead and [refactored entirely according to our discussion](https://github.com/Akrabut/bundler/pull/4/files) (except for the tests).
Would love to hear your opinion - does it look better than this version?

Reply:
Yes, I think it's better. 👍

Reply:
Thanks, I'll wait for 1-2 more opinions from the core team and I'll replace this PR with the new one.